packages/core/src/routes/interaction/verifications/mfa-verification.ts

```diff
@@ -73,6 +74,77 @@ export const verifyMfa = async (
   return interaction;
 };
 
+const userMfaDataKey = 'mfa';
+/**
+ * Check if the user has skipped MFA binding
+ */
+const isMfaSkipped = (customData: JsonObject): boolean => {
+  const userMfaDataGuard = z.object({
+    skipped: z.boolean().optional(),
+  });
+
+  const parsed = z.object({ [userMfaDataKey]: userMfaDataGuard }).safeParse(customData);
+
+  return parsed.success ? parsed.data[userMfaDataKey].skipped === true : false;
+};
+
+const validateMandatoryBindMfaForSignIn = async (
+  tenant: TenantContext,
+  ctx: WithInteractionSieContext & WithInteractionDetailsContext,
+  interaction: VerifiedSignInInteractionResult
+): Promise<VerifiedInteractionResult> => {
+  const {
+    mfa: { policy, factors },
+  } = ctx.signInExperience;
+  const { bindMfas } = interaction;
+  const availableFactors = factors.filter((factor) => factor !== MfaFactor.BackupCode);
+
+  // No available MFA, skip check
+  if (availableFactors.length === 0) {
+    return interaction;
+  }
+
+  const hasFactorInBind = Boolean(
+    bindMfas &&
+      availableFactors.some((factor) => bindMfas.some((bindMfa) => bindMfa.type === factor))
+  );
+
+  const { accountId } = interaction;
+  const { mfaVerifications, customData } = await tenant.queries.users.findUserById(accountId);
+  const hasFactorInUser = factors.some((factor) =>
+    mfaVerifications.some(({ type }) => type === factor)
+  );
+
+  if (hasFactorInBind || hasFactorInUser) {
```

Comment:
wrap this check into a function? Or add some comments to clarify

Reply:
comments added